Vendor Catch2: v2.5.0

ax3l opened the pull request:
Add a third-party header for a vendored catch2 for our tests.

Source: https://github.com/catchorg/Catch2/releases
Version: 0.2.5

## New CMake Option

By default, we will build the tests with the internally vendored Catch2 in order to make testing easy. If you want to force using an external Catch2 install (expand your `CMAKE_PREFIX_PATH` with its location), then set: `-DALPAKA_USE_INTERNAL_CATCH2=OFF`

ax3l committed:
CMake: Control Internal/External Catch2

ax3l committed:
CI: Vendored Catch2

Use the vendored Catch2 in Alpaka instead of installing an
external one.

BenjaminW3 commented on script/travis/before_install.sh:
```diff
@@ -89,7 +89,6 @@ if [ "${ALPAKA_ACC_GPU_CUDA_ENABLE}" == "ON" ] || [ "${ALPAKA_ACC_GPU_HIP_ENABLE
 then
     if [ "${ALPAKA_CUDA_COMPILER}" == "nvcc" ]
     then
```
Please try to remove the whole `CUDA` block.

philsquared committed:
Third Party: Vendor Catch 2.5.0

Add the single-header release of Catch 2.5.0 to a third party
library for tests.

BenjaminW3 commented:
The changes hit a known bug in gcc-5.5:
* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=76731
* https://github.com/tensorflow/tensorflow/issues/10220 .

It works with gcc-5.4, gcc-4.9 or gcc-6 and also when building with nvcc 10.0. It previously worked because we had been building debug builds only but not release.

Could you please revert the changes in `before_install.sh` so that this can be merged and create a bug ticket to resolve the nvcc-9+gcc-5.5+release issue?

BenjaminW3 commented on test/CMakeLists.txt:
```diff
@@ -26,9 +26,15 @@ FIND_PACKAGE(alpaka REQUIRED)
-message(STATUS "Catch2: Found version ${Catch2_VERSION}")
+OPTION(ALPAKA_USE_INTERNAL_CATCH "Use internally shipped Catch2" ON)
+
+IF(ALPAKA_USE_INTERNAL_CATCH)
```
Could we rename it to `ALPAKA_USE_INTERNAL_CATCH2`?

ax3l replied:
sure!
